### PR TITLE
avm2: Implement JSON

### PR DESCRIPTION
**Pull request by EmperorBale**
Implements the JSON class for serializing and deserializing AVM and JSON values.

- [x] `JSON.parse`
- [x] `JSON.stringify`
- [x] Write tests

### CODE REVIEW

**Comment by EmperorBale** [comment]
Going to finish this PR now

**Comment by EmperorBale** [comment]
So writing the test for JSON.stringify was a bit tricky because the order of elements that are added to the json string is random, and will be different each time. So instead of testing for whether the output strings are the same, I decided to test if the length is the same instead since that will always be the same, regardless of order.

File: core/src/avm2/globals/json.rs [hunk 1]
```diff
@@ -0,0 +1,443 @@
...
+                num.into()
+            }
+        }
+        JsonValue::Object(js_obj) => {
+            let obj_class = activation.avm2().classes().object;
+            let mut obj = obj_class.construct(activation, &[])?;
+            for entry in js_obj.iter() {
+                let key = AvmString::new_utf8(activation.context.gc_context, entry.0);
+                let val = deserialize_json_inner(activation, entry.1.clone(), reviver)?;
+                let mapped_val = reviver.map_or(Ok(val.clone()), |reviver| {
```
Comment (adrian17) [inline core/src/avm2/globals/json.rs:48]: I think that if you replaced `map_or` by `map_or_else` or explicit match, then both branches would directly consume `val`, so you could avoid some cloning.

File: core/src/avm2/globals/json.rs [hunk 2]
```diff
@@ -0,0 +1,443 @@
...
+                    reviver.call(None, &[key.into(), val], activation)
+                })?;
+                if matches!(mapped_val, Value::Undefined) {
+                    obj.delete_property(activation, &QName::new(Namespace::public(), key).into())?;
+                } else {
+                    obj.set_property(
+                        obj,
+                        &QName::new(Namespace::public(), key).into(),
```
Comment (adrian17) [inline core/src/avm2/globals/json.rs:56]: FYI we have `QName::dynamic_name(key)` as a shorthand.
(I may refactor the naming in the future :D )

File: core/src/avm2/globals/json.rs [hunk 3]
```diff
@@ -0,0 +1,443 @@
...
+    })
+}
+
+fn deserialize_json<'gc>(
+    activation: &mut Activation<'_, 'gc, '_>,
+    json: JsonValue,
+    reviver: Option<Object<'gc>>,
+) -> Result<Value<'gc>, Error> {
+    let val = deserialize_json_inner(activation, json, reviver)?;
+    reviver.map_or(Ok(val.clone()), |reviver| {
```
Comment (adrian17) [inline core/src/avm2/globals/json.rs:86]: same here about the `clone()`.

File: core/src/avm2/globals/json.rs [hunk 4]
```diff
@@ -0,0 +1,443 @@
...
+}
+
+enum Replacer<'gc> {
+    Function(FunctionObject<'gc>),
+    PropList(ArrayObject<'gc>),
+}
+
+struct AvmSerializer<'gc> {
+    /// This object stack will be used to detect circular references and return an error instead of a panic.
+    obj_stack: Vec<*const ObjectPtr>,
```
Comment (adrian17) [inline core/src/avm2/globals/json.rs:164]: I've seen this was a simple `Object<'gc>` before, so I'm curious - why the change to pointer?

File: core/src/avm2/globals/json.rs [hunk 5]
```diff
@@ -0,0 +1,443 @@
...
+                if !matches!(mapped, Value::Undefined) {
+                    js_obj.insert(
+                        &key.to_utf8_lossy(),
+                        self.serialize_value(activation, mapped)?,
+                    );
+                }
+            }
+        } else {
+            for i in 1.. {
+                // TODO: We should get more than just enumerable properties
```
Comment (adrian17) [inline core/src/avm2/globals/json.rs:251]: Some extra "fun" cases:
stringifying a function appears to produce `null`
stringifying a class object 
```
class C {
    public static var X = 1;
}
```
appears to produce `{"X":1,"prototype":{}}`

Also, in general, it does indeed iterate more than just enumerable properties.
```
dynamic class C {
    public var X = 1;
    public var Y = 2;
}
var a = new C();
a.Z = 3;

for (var x in a) {
    trace(x);
}

trace(JSON.stringify(a));
```
produces
```
Z
{"X":1,"Y":2,"Z":3}
```

BTW this might be more consistent to write after my refactor PR, as I introduce a clean split between non-dynamic (trait) and dynamic instance properties. But I think it'd also me okay if this got merged first and improved later, as long as we have tests ready.

File: core/src/avm2/globals/json.rs (hunk 4 above)
Comment (EmperorBale) [reply core/src/avm2/globals/json.rs:164]: The reason was because I assumed that storing the pointer would be cheaper than storing the whole object (which contains extra metadata). Although, now that I think about it, the call to `as_ptr()` may be more expensive than I thought because it needs to use the objects vtable, so I will probably revert this change

Comment (adrian17) [reply core/src/avm2/globals/json.rs]: I just tested it on desktop, and actually `obj1 == obj2` also compiles terribly due to virtual call :( The approach with `*const ObjectPtr` avoids _some_ repeated virtual calls, so maybe let's stay with that for now? But it's not that perf-critical 
 either and might be improved in the future too... up to you :)
